records/2021-09-27.batch-celery-save.md

```diff
@@ -0,0 +1,116 @@
+in the current task. Alternatively, if only 2,300 requests filled up and the 5
+seconds timeout completed, then all of the 2,3000 requests would get flushed to
+the task.
+
+This is a transparent and easy mechanism that could get leveraged in Celery in order
+to achieve the batch save. Unfortunately, that feature was always in an experimental
+state in Celery version 3 and disappeared in subsequent versions. An individual
+contributer on GitHub [made a fork of the experimental feature](https://github.com/clokep/celery-batches) and adapted for recent
+versions of Celery. This looked promising unfortunately there are two issues by
+using this library:
+
+1- It still can be considered into an experimental stage as this does not supports common basic feature offered by Celery such [as a retry mechanism](https://github.com/clokep/celery-batches/issues/10).
+2- The integration is not that easy with Flask (although probably not impossible) as described by Flask integration recipe [in their documentation](https://flask.palletsprojects.com/en/2.0.x/appcontext/). [An issue was opened](https://github.com/clokep/celery-batches/issues/35) for curiosity sake to see if that could be done but even then, the issue described at #1 is a blocker.
+
+### Send notifications in queue and sweep process from a heartbeat process
+
+> **INFO**: There is a [Hearbeat](https://docs.celeryproject.org/en/stable/internals/reference/celery.worker.heartbeat.html) service within Celery that the latter manages. This is a one and only execution thread which allows to orchestrate certain tasks based on a schedule. This is useful as we can leverage a memory model that we will not share. Hence we avoid the pitfalls of concurrent execution thread trying to access shared memory all at the same time and which requires complex locking mechanism. This is similar to the UI model with an unique UI thread and its exclusive access to UI data, or the actor model which has one actor keeping its one exclusive state kept away from other actors.
+
+In order to achieve batch save, we need to buffer and retrieve the items from the
+Celery task. There is unfortunately no known mechanism at this time of writing of
+Celery that allows the retrieval of all items for a queue and process it as a batch
+(other than the previously mentioned and defunct celery-batch experimental library).
+
+There are [Celery primitives that allows to chain, iterate concurrently and group](https://docs.celeryproject.org/en/stable/userguide/canvas.html#the-primitives)
+which will be necessary to achieve a custom solution but no aggregate operation is
+available.
+
+Hence, we will need to buffer our notifications into a queue or cache system when
+we receive the notifications, prior to save these.
+
+![Buffer workflow of a notification save](https://raw.githubusercontent.com/cds-snc/notification-adr/main/records/diagrams/2021-09-27.batch-celery-save/buffer-workflow.jpg?raw=true "Buffer workflow of a notification save")
```

Comment:
I added a legend and reverted queue names in that diagram as well, per @dj2 's comments on the first diagram.

Reply:
So the new idea here is that rather than write the new notifications into RDS one at a time, we send them into a queue, and the `bulk_save_email` task pulls out a batch of them and writes them in one operations?

Reply:
Yes! we want to batch insert these.